### PR TITLE
Use HRESULT instead of bool

### PR DESCRIPTION
**Pull request by leculver**
- Intentionally start accepting more SUCCEEDED(hr) instead of hr == S_OK.
- Created an HResult struct to wrap COM's HRESULT.
- HResult has an implicit int cast operator for convenience.
- HResult has an implicit bool cast operator which is the equivalent of "SUCCEEDED(hr)".
- Change most wrappers to use and return HResult instead of bool.
- Functionality should be mostly unchanged, but allows for better control/error flow.

### CODE REVIEW

**Comment by leculver** [comment]
This is complete, but awaiting work to un-break our internal arcade build system which I'm working on now.

**Comment by leculver** [comment]
> There is a comment on L33.

L33 of what file?

> Is this a bug? I haven't seen anything fail before.

Not a bugfix, just general cleanup to more closely align with how I handle COM interfaces in other (internal) projects I work on.  I prefer to have the HRESULT around somewhere rather than a true/false when debugging failures.

File: src/Microsoft.Diagnostics.Runtime/src/DbgEng/DebugSystemObjects.cs [hunk 1]
```diff
@@ -29,41 +29,45 @@ public uint GetProcessId()
...
-            DebugOnly.Assert(hr == S_OK);
+            HResult hr = _getNumberThreads(Self, out int count);
+            DebugOnly.Assert(hr);
             return count;
         }
 
         internal void Init()
         {
             InitDelegate(ref _getSystemId, VTable.GetCurrentSystemId);
-            _systemId = _getSystemId(Self, out int id);
```
Comment (leculver) [inline src/Microsoft.Diagnostics.Runtime/src/DbgEng/DebugSystemObjects.cs:66]: @NextTurn Actually there was one bug "fixed" with this checkin, it's this line.  We originally set the _systemId to be the HRESULT of the return of `IDebugSystemObjects::GetSystemId`, which is a bug.

I fixed that here, but I'm seeing odd behavior in tests, so I'm debugging the problem now.

File: src/Microsoft.Diagnostics.Runtime/src/DacInterface/ClrDataProcess.cs [hunk 2]
```diff
@@ -118,13 +109,12 @@ public IEnumerable<ClrDataMethod> EnumerateMethodInstancesByAddress(ClrDataAddre
...
 
-            int hr = _startEnum(Self, addr, IntPtr.Zero, out ClrDataAddress handle);
-            if (hr != S_OK)
+            if (!_startEnum(Self, addr, IntPtr.Zero, out ClrDataAddress handle))
                 return result;
 
             try
             {
-                while ((hr = _enum(Self, ref handle, out IntPtr method)) == S_OK)
+                while (_enum(Self, ref handle, out IntPtr method))
```
Comment (dailongchen) [inline src/Microsoft.Diagnostics.Runtime/src/DacInterface/ClrDataProcess.cs:117]: HResult.S_OK is zero, this change breaks this function.

Comment (NextTurn) [reply src/Microsoft.Diagnostics.Runtime/src/DacInterface/ClrDataProcess.cs:117]: Anything wrong?

#### Before

`hr == 0`

#### After

`hr >= 0`

Comment (leculver) [reply src/Microsoft.Diagnostics.Runtime/src/DacInterface/ClrDataProcess.cs:117]: Yep that's a bug.  Thanks!  I've fixed this here, which will be released in a nuget package this week:  https://github.com/microsoft/clrmd/pull/819

**Pull request merged**